include/seqan3/search/algorithm/search_result.hpp

```diff
@@ -0,0 +1,104 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::search_result.
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/core/type_traits/template_inspection.hpp>
+
+namespace seqan3::detail
+{
+
+template <typename iterator_t,
+          typename sequence_id_t = std::nullopt_t *,
+          typename sequence_pos_t = std::nullopt_t *>
+struct search_result_value_type
+{
+    size_t query_id{};
+    iterator_t iterator{};
+    sequence_id_t sequence_id{};
+    sequence_pos_t sequence_pos{};
+
+    search_result_value_type() = default;                                                //!< Defaulted
+    search_result_value_type(search_result_value_type const &) = default;                //!< Defaulted
+    search_result_value_type(search_result_value_type &&) = default;                     //!< Defaulted
+    search_result_value_type & operator=(search_result_value_type const &) = default;    //!< Defaulted
+    search_result_value_type & operator=(search_result_value_type &&) = default;         //!< Defaulted
+    ~search_result_value_type() = default;                                               //!< Defaulted
+
+    // No position.
+    search_result_value_type(size_t _id, iterator_t _it) :
+        query_id(_id), iterator(_it) {}
+
+    // Only one size_t that indicates sequence position. sequence id is then 0.
+    search_result_value_type(size_t _id, iterator_t _it, sequence_pos_t _pos) :
+    query_id(_id), iterator(_it), sequence_id(0), sequence_pos(_pos) {}
+
+    // Search returns a pair of sequence if and position.
+    search_result_value_type(size_t _id, iterator_t _it, std::pair<sequence_id_t, sequence_pos_t> _pair) :
+        query_id(_id), iterator(_it), sequence_id(_pair.first), sequence_pos(_pair.second) {}
+
+
+};
+
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+
+template <typename search_result_traits>
+    requires detail::is_type_specialisation_of_v<search_result_traits, detail::search_result_value_type>
+class search_result
+{
+private:
+    search_result_traits data;
+
+    using iterator_t          = decltype(data.iterator);
```

Comment:
I would prefer not aligning the stuff. 
I mean if the documentation will be added this won't matter anyway so much